### ISSUE TITLE
kubectl create -f - doesn't respect EOF

### ISSUE THREAD

**Issue by thockin** [kubernetes/kubernetes#34345]
When running `kubectl create -f -` it reads stdin, but when I hit ^D (EOF) it does not stop reading input.

**Comment by PhilibertDugas** [comment]
I'm thinking this line is at fault: https://github.com/kubernetes/kubernetes/blob/8fd414537b5143ab039cb910590237cabf4af783/pkg/util/yaml/decoder.go#L67

If `err == io.EOF` it doesn't return it, even tough the `StreamVisitor` handles it:
https://github.com/kubernetes/kubernetes/blob/6f6929324020060d77a36e7c5de691d25334a8bc/pkg/kubectl/resource/visitor.go#L525

My first reflex would be to change the condition in the decoder, but I'm not sure of the potential side effects. Thoughts?

**Comment by thockin** [comment]
Try it out?

**Comment by PhilibertDugas** [comment]
I'll take a stab at it during the weekend

**Comment by AdoHe** [comment]
@PhilibertDugas the problem is that: get `io.EOF` but still get some content. In this case, we will not return err.

**Comment by PhilibertDugas** [comment]
@AdoHe I had found something else, let me know if this makes sense:

To decode YAML or JSON, the decoder tries to `GuessJSONStream`: https://github.com/kubernetes/kubernetes/blob/master/pkg/util/yaml/decoder.go#L201

This method does a `Peek` call:
https://github.com/kubernetes/kubernetes/blob/master/pkg/util/yaml/decoder.go#L289

The problem with `Peek` and the `os.Stdin` stream is that it will block until an `io.EOF` character is sent, therefore the whole method call is blocked.  I found a description of the issue here: http://stackoverflow.com/questions/22744443/check-if-there-is-something-to-read-on-stdin-in-golang

The behaviour I see is that the first time I press `^D` the `Seek` calls returns. At this point, we see this log return: https://github.com/kubernetes/kubernetes/blob/master/pkg/util/yaml/decoder.go#L206

The second time I press `^D` the line is actually read, in this case I have an error because I didn't enter anything.

![ioeof](https://cloud.githubusercontent.com/assets/5460957/19216656/0c016b7e-8d93-11e6-9aad-95e440c29f4e.gif)

I'm thinking we could save the bytes returned from the `Peek` method if we know that we have a terminal `Stdin` and read from those bytes.

**Comment by AdoHe** [comment]
@PhilibertDugas would you mind first give a try with kubernetes 1.3? I want to make sure whether this is a regression or not, as we changed decoder in 1.4 to replace the scanner based approach.

**Comment by PhilibertDugas** [comment]
Ah yes, sorry for not sharing this, I believe my `kubectl` client is on 1.3 already:

![image](https://cloud.githubusercontent.com/assets/5460957/19216705/68614180-8d95-11e6-98ce-9937fc2bb6a2.png)

**Comment by ronaldpetty** [comment]
I am using 1.6.0-alpha and ctrl-d seems to work.  Since I am new, how can I help close this one?  Adding a test to exercise?

```
user@ubuntu:~/kubernetes$ cluster/kubectl.sh create -f -
apiVersion: v1
kind: Pod
metadata:
  name: nginxpod
spec:
  containers:
  - name: nginx
    image: nginx
    ports:
    - containerPort: 80
pod "nginxpod" created
user@ubuntu:~/kubernetes$ cluster/kubectl.sh get pods
NAME       READY     STATUS              RESTARTS   AGE
nginxpod   0/1       ContainerCreating   0          12s
```

**Comment by xiangpengzhao** [comment]
I reproduce the issue on v1.5.1 but I don't reproduce it on HEAD. So I think this was fixed in some version later than v1.5.1.

@thockin can we close this issue then?
